ext/ddtrace.c

```diff
@@ -621,6 +621,117 @@ static PHP_GSHUTDOWN_FUNCTION(ddtrace) {
+    ZVAL_UNDEF(&garbage_attributes);
+
+    // Copy current values to temporary zval variables
+    ZVAL_COPY_VALUE(&garbage_name, &event->property_name);
+    ZVAL_COPY_VALUE(&garbage_attributes, &event->property_attributes);
+
+    // Update event properties
+    ZVAL_STR_COPY(&event->property_name, name);
+
+    // Use the provided timestamp or the current time in nanoseconds
+    if (timestamp == 0) {
+        struct timespec ts;
+        timespec_get(&ts, TIME_UTC);
+        ZVAL_LONG(&event->property_timestamp, ts.tv_sec * ZEND_NANO_IN_SEC + ts.tv_nsec);
+    } else {
+        ZVAL_LONG(&event->property_timestamp, timestamp);
+    }
+
+    if (attributes && Z_TYPE_P(attributes) == IS_ARRAY) {
+        ZVAL_COPY(&event->property_attributes, attributes);
+    } else {
+        array_init(&event->property_attributes);
+    }
+
+    // Free the copied values after replacement
+    zval_ptr_dtor(&garbage_name);
+    zval_ptr_dtor(&garbage_attributes);
+}
+
+/* DDTrace\SpanEvent */
+zend_class_entry *ddtrace_ce_span_event;
+
+PHP_METHOD(DDTrace_SpanEvent, jsonSerialize) {
+    ddtrace_span_event *event = (ddtrace_span_event*)Z_OBJ_P(ZEND_THIS);
+
+    zend_array *array = zend_new_array(3);
+
+    zend_string *name = zend_string_init("name", sizeof("name") - 1, 0);
+    zend_string *timestamp = zend_string_init("time_unix_nano", sizeof("time_unix_nano") - 1, 0);
+
+    Z_TRY_ADDREF(event->property_name);
+    zend_hash_add(array, name, &event->property_name);
+    Z_TRY_ADDREF(event->property_timestamp);
+    zend_hash_add(array, timestamp, &event->property_timestamp);
+
+    zend_string_release(name);
+    zend_string_release(timestamp);
+
+    if (Z_TYPE(event->property_attributes) == IS_ARRAY && Z_ARRVAL(event->property_attributes)->nNumOfElements > 0) {
+        zend_string *attributes = zend_string_init("attributes", sizeof("attributes") - 1, 0);
+        Z_TRY_ADDREF(event->property_attributes);
+        zend_hash_add(array, attributes, &event->property_attributes);
+        zend_string_release(attributes);
+    }
+
+    RETURN_ARR(array);
+}
+
+PHP_METHOD(DDTrace_SpanEvent, __construct)
+{
+    UNUSED(return_value);
+
+    zend_string *name;
+    zval *attributes = NULL;
+    zend_long timestamp = 0;
+
+    ZEND_PARSE_PARAMETERS_START(1, 3)
+        Z_PARAM_STR(name)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_ARRAY_EX(attributes, 1, 0)
+        Z_PARAM_LONG(timestamp)
+    ZEND_PARSE_PARAMETERS_END();
+
+    ddtrace_span_event *event = (ddtrace_span_event*)Z_OBJ_P(ZEND_THIS);
+
+    // Use the static function to set properties and handle cleanup
+    dd_span_event_construct(event, name, attributes, timestamp);
+}
+
+/* DDTrace\ExceptionSpanEvent */
+zend_class_entry *ddtrace_ce_exception_span_event;
+
+PHP_METHOD(DDTrace_ExceptionSpanEvent, __construct)
+{
+    UNUSED(return_value);
+
+    zval *exception;
+    zval *attributes = NULL;
+
+    ZEND_PARSE_PARAMETERS_START(1, 2)
+        Z_PARAM_OBJECT_OF_CLASS(exception, zend_ce_throwable)
```

Comment:
Parameter is still unused.

Reply:
On the latest commit moved the code for the attribute handling in here instead. Let me know what you think of that please.

Reply:
Can you move the attribute handling to the jsonSerialize function instead? So that changes to the exception property after object construction get reflected there.

Reply:
So I've been trying saving it to the ExceptionSpanEvent and then in the serializer make use of that but then my `phpt` fail where my `Interopability` one doesn't, not sure if you can spot why or provide any suggestions: https://github.com/DataDog/dd-trace-php/pull/2754/commits/ef1341aa1a84643d6f830d7fa2bd2bedf0dcfc72